cni/pkg/nftables/nftables.go

```diff
@@ -0,0 +1,769 @@
+		)
+
+		// CLI: nft add rule inet istio-ambient-raw istio-prerouting udp sport 53 meta mark and 0xfff != 0x539 counter ct zone set 1
+		// Upstream --> Proxy return packets
+		cfg.ruleBuilder.AppendRule(
+			IstioPreroutingChain, AmbientRawTable,
+			"udp sport", "53",
+			"meta mark and", fmt.Sprintf("0x%x", config.InpodMask),
+			"!=", fmt.Sprintf("0x%x", config.InpodMark), Counter,
+			"ct zone set", "1",
+		)
+	}
+
+	// CLI: nft add rule inet istio-ambient-nat istio-output meta l4proto tcp mark and 0xfff == 0x111 counter accept
+	//
+	// DESC: If this is outbound and has our mark, let it go.
+	cfg.ruleBuilder.AppendRule(
+		IstioOutputChain, AmbientNatTable,
+		"meta l4proto tcp",
+		"mark and", fmt.Sprintf("0x%x", config.InpodTProxyMask),
+		"==", fmt.Sprintf("0x%x", config.InpodTProxyMark), Counter,
+		"accept",
+	)
+
+	// Do not redirect app calls to back itself via Ztunnel when using the endpoint address
+	// e.g. appN => appN by lo
+	// CLI: nft add rule inet istio-ambient-nat istio-output oifname "lo" ip daddr != 127.0.0.1 counter accept
+	cfg.ruleBuilder.AppendRule(
+		IstioOutputChain, AmbientNatTable,
+		"oifname", "lo",
+		"ip daddr",
+		"!=", "127.0.0.1/32", Counter,
+		"accept",
+	)
+
+	// CLI: nft add rule inet istio-ambient-nat istio-output oifname "lo" ip6 daddr != ::1/128 counter accept
+	cfg.ruleBuilder.AppendRule(
+		IstioOutputChain, AmbientNatTable,
+		"oifname", "lo",
+		"ip6 daddr",
+		"!=", "::1/128", Counter,
+		"accept",
+	)
+
+	// CLI: nft add rule inet istio-ambient-nat istio-output meta l4proto tcp ip daddr != 127.0.0.1 mark and 0xfff != 0x539 counter redirect to :15001
+	//
+	// DESC: If this is outbound, not bound for localhost, and does not have our packet mark, redirect to ztunnel proxy <OUTPORT>
+	cfg.ruleBuilder.AppendRule(
+		IstioOutputChain, AmbientNatTable,
+		"meta l4proto tcp",
+		"ip daddr",
+		"!=", "127.0.0.1/32",
+		"mark and", fmt.Sprintf("0x%x", config.InpodMask),
+		"!=", fmt.Sprintf("0x%x", config.InpodMark), Counter,
+		"redirect to", ":"+fmt.Sprintf("%d", config.ZtunnelOutboundPort),
+	)
+
+	cfg.ruleBuilder.AppendRule(
+		IstioOutputChain, AmbientNatTable,
+		"meta l4proto tcp",
+		"ip6 daddr",
+		"!=", "::1/128",
+		"mark and", fmt.Sprintf("0x%x", config.InpodMask),
+		"!=", fmt.Sprintf("0x%x", config.InpodMark), Counter,
+		"redirect to", ":"+fmt.Sprintf("%d", config.ZtunnelOutboundPort),
+	)
+
+	return cfg.executeCommands()
+}
+
+// DeleteInpodRules removes nftables rules from a pod's network namespace
+func (cfg *NftablesConfigurator) DeleteInpodRules(log *istiolog.Scope) error {
+	log.Info("removing nftables inpod rules")
+
+	nft, err := cfg.nftProvider("", "")
+	if err != nil {
+		return err
+	}
+
+	tx := nft.NewTransaction()
+	// nftables delete command will delete the table along with the chains and rules.
+	tx.Add(&knftables.Table{Name: AmbientNatTable, Family: knftables.InetFamily})
+	tx.Delete(&knftables.Table{Name: AmbientNatTable, Family: knftables.InetFamily})
```

Comment:
also why Delete and not Flush?

Reply:
Since flushing only removes the rules from the chains in a table, while deleting removes the entire table along with its chains, and this API is called when a Pod is removed from the Ambient mesh, it’s better to return to a clean state by deleting the table and its associated chains.